velox/dwio/parquet/tests/ParquetTpchTest.cpp

```diff
@@ -0,0 +1,221 @@
+          {VARCHAR(),
+           VARCHAR(),
+           DOUBLE(),
+           DOUBLE(),
+           DOUBLE(),
+           DOUBLE(),
+           VARCHAR()});
+
+  // date '1998-12-01' - interval '90' day <= shipdate <= '1998-12-01'
+  auto filters =
+      common::test::SubfieldFiltersBuilder()
+          .add("shipdate", common::test::between("1998-09-02", "1998-12-01"))
+          .build();
+
+  const int sourcePlanNodeId = 10;
+  static const core::SortOrder kAscNullsLast(true, false);
+
+  const auto stage1 = PlanBuilder(sourcePlanNodeId)
+                          .tableScan(
+                              rowType,
+                              makeTableHandle(std::move(filters)),
+                              allRegularColumns(rowType))
+                          .project(
+                              {"returnflag",
+                               "linestatus",
+                               "quantity",
+                               "extendedprice",
+                               "extendedprice * (1.0 - discount)",
+                               "extendedprice * (1.0 - discount) * (1.0 + tax)",
+                               "discount"})
+                          .partialAggregation(
+                              {0, 1},
+                              {"sum(quantity)",
+                               "sum(extendedprice)",
+                               "sum(p4)",
+                               "sum(p5)",
+                               "avg(quantity)",
+                               "avg(extendedprice)",
+                               "avg(discount)",
+                               "count(0)"})
+                          .planNode();
+
+  auto plan = PlanBuilder(1)
+                  .localPartition({0, 1}, {stage1})
+                  .finalAggregation(
+                      {0, 1},
+                      {"sum(a0)",
+                       "sum(a1)",
+                       "sum(a2)",
+                       "sum(a3)",
+                       "avg(a4)",
+                       "avg(a5)",
+                       "avg(a6)",
+                       "count(a7)"},
+                      {DOUBLE(),
+                       DOUBLE(),
+                       DOUBLE(),
+                       DOUBLE(),
+                       DOUBLE(),
+                       DOUBLE(),
+                       DOUBLE(),
+                       BIGINT()})
+                  .orderBy({0, 1}, {kAscNullsLast, kAscNullsLast}, false)
+                  .planNode();
+
+  CursorParameters params;
+  params.planNode = plan;
+  params.maxDrivers = 4;
+  params.numResultDrivers = 1;
+
+  bool noMoreSplits = false;
+  auto result = readCursor(params, [&](auto* task) {
+    addSplitsToTask(task, noMoreSplits, filePaths, sourcePlanNodeId);
+  });
+
+  EXPECT_EQ(1, result.second.size());
+  auto output = materialize(result.second[0]);
+  ASSERT_EQ(10, output[0].size());
+  ASSERT_EQ(std::string("N"), std::string(output[0][0].value<StringView>()));
```

Comment:
These assertions are pretty hard to read. Consider using assertQuery to compare the results with DuckDB. You can use TPC-H extension provided with DuckDB.

Reply:
The DuckDB TPC-H extension has only a single API to get results
https://github.com/duckdb/duckdb/blob/01bdd7ce8930a960ad4e9ded58f29efa8fb65cda/extension/tpch/include/tpch-extension.hpp#L23
```
//! Returns the CSV answer of a TPC-H query
static std::string GetAnswer(double sf, int query);
```
Instead of parsing CSV, the current change directly queries the Parquet files using the Parquet extension.
This does mean an additional step to copy the `parquet-extension.hpp` file similar to TPC-H extension.
Any thoughts?

Reply:
@Mytherin @hannesmuehleisen What would be a proper way to run a TPC-H query in DuckDB and get results as a vector, not CSV?

Reply:
@majetideepak Have you tried fetching query test using "static std::string GetQuery(int query);" API and running it the same way we run other queries, e.g. via assertQuery?

Reply:
@mbasmanova `::duckdb::TPCHExtension::GetQuery(1)` returns the following string. 
```
"SELECT\n    l_returnflag,\n    l_linestatus,\n    sum(l_quantity) AS sum_qty,\n    sum(l_extendedprice) AS sum_base_price,\n    sum(l_extendedprice * (1 - l_discount)) AS sum_disc_price,\n    sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)) AS sum_charge,\n    avg(l_quantity) AS avg_qty,\n    avg(l_extendedprice) AS avg_price,\n    avg(l_discount) AS avg_disc,\n    count(*) AS count_order\nFROM\n    lineitem\nWHERE\n    l_shipdate <= CAST('1998-09-02' AS date)\nGROUP BY\n    l_returnflag,\n    l_linestatus\nORDER BY\n    l_returnflag,\n    l_linestatus;\n"
```
We can use this only after creating a duckdb table `ineitem` first correct? That means we will have to create the tpc-h `lineitem` data manually and pass it to `DuckDbQueryRunner::createTable(lineitem, data)`. This is a redundant step correct right?

Did I miss something?

Reply:
Alternatively, we could add `DuckDbQueryRunner::createTableFromParquet(lineitem, filePaths)` and implement it as
```
CREATE TABLE lineitem AS SELECT * FROM parquet_scan('');
```
But we still need to add the Parquet extension for this to work.

Reply:
The other issue with the current `assertQuery` API is that there is no way to specify `params.maxDrivers = 4;` and the source plan id must always be `0`

Reply:
I figured out how to load the TPCH data into DuckDB. We can invoke `::duckdb::TPCHExtension::GetQuery(1)` and should be good.
```
    void initTPCH() {
      db_.LoadExtension<::duckdb::TPCHExtension>();
      ::duckdb::Connection con(db_);
      con.Query("CALL dbgen(sf=0.01)");
    }
```

Reply:
That's great.

Reply:
> The other issue with the current `assertQuery` API is that there is no way to specify `params.maxDrivers = 4;` and the source plan id must always be `0`

There is a version of assertQuery that takes CursorParameters and addSplits lambda in velox/exec/tests/utils/QueryAssertions.h. This allows you to specify params.maxDrivers and use whatever plan node ID you want when adding splits.

```
std::shared_ptr<Task> assertQuery(
    const CursorParameters& params,
    std::function<void(exec::Task*)> addSplits,
    const std::string& duckDbSql,
    DuckDbQueryRunner& duckDbQueryRunner,
    std::optional<std::vector<uint32_t>> sortingKeys = std::nullopt);
```

Reply:
Ideally, we would use DuckDB's Parquet writer to generate Parquet files on the fly instead of checking them in.

Reply:
Makes sense. I will try generating the files.